go/internal/feast/featurestore.go

```diff
@@ -0,0 +1,885 @@
+		}
+
+		for featureName, occurrences := range collidedFeatureRefs {
+			if occurrences == 1 {
+				delete(collidedFeatureRefs, featureName)
+			}
+		}
+		if len(collidedFeatureRefs) >= 1 {
+			collidedFeatureRefList := make([]string, 0)
+			for _, featureRef := range featureRefs {
+				_, featureName, err := parseFeatureReference(featureRef)
+				if err != nil {
+					return err
+				}
+				if _, ok := collidedFeatureRefs[featureName]; ok {
+					collidedFeatureRefList = append(collidedFeatureRefList, featureRef)
+				}
+
+			}
+			return errors.New(fmt.Sprintf("featureNameCollisionError: %s; %t", strings.Join(collidedFeatureRefList, ", "), fullFeatureNames))
+		}
+	}
+	return nil
+}
+
+func (fs *FeatureStore) getNeededRequestData(groupedOdfvRefs map[*OnDemandFeatureView][]string,
+	groupedRequestFvRefs map[*RequestFeatureView][]string) (map[string]struct{}, map[string]struct{}, error) {
+	neededRequestData := make(map[string]struct{})
+	neededRequestFvFeatures := make(map[string]struct{})
+
+	for onDemandFeatureView := range groupedOdfvRefs {
+		requestSchema := onDemandFeatureView.getRequestDataSchema()
+		for fieldName := range requestSchema {
+			neededRequestData[fieldName] = struct{}{}
+		}
+	}
+
+	for requestFeatureView := range groupedRequestFvRefs {
+		for _, feature := range requestFeatureView.base.features {
+			neededRequestFvFeatures[feature.name] = struct{}{}
+		}
+	}
+
+	return neededRequestData, neededRequestFvFeatures, nil
+}
+
+func (fs *FeatureStore) ensureRequestedDataExist(neededRequestData map[string]struct{},
+	neededRequestFvFeatures map[string]struct{},
+	requestDataFeatures map[string]*types.RepeatedValue) error {
+	// TODO (Ly): Review: Skip checking even if composite set of
+	// neededRequestData neededRequestFvFeatures is different from
+	// request_data_features but same length?
+	if len(neededRequestData)+len(neededRequestFvFeatures) != len(requestDataFeatures) {
+		missingFeatures := make([]string, 0)
+		for feature := range neededRequestData {
+			if _, ok := requestDataFeatures[feature]; !ok {
+				missingFeatures = append(missingFeatures, feature)
+			}
+		}
+		for feature := range neededRequestFvFeatures {
+			if _, ok := requestDataFeatures[feature]; !ok {
+				missingFeatures = append(missingFeatures, feature)
+			}
+		}
+		return errors.New(fmt.Sprintf("requestDataNotFoundInEntityRowsException: %s", strings.Join(missingFeatures, ", ")))
+	}
+	return nil
+}
+
+func (fs *FeatureStore) checkOutsideTtl(featureTimestamp *timestamppb.Timestamp, currentTimestamp *timestamppb.Timestamp, ttl *durationpb.Duration) bool {
+	return currentTimestamp.GetSeconds()-featureTimestamp.GetSeconds() > ttl.Seconds
+}
+
+func (fs *FeatureStore) populateResultRowsFromColumnar(response *serving.GetOnlineFeaturesResponse, data map[string]*types.RepeatedValue) {
+	timeStamp := timestamppb.Now()
+	for entityName, values := range data {
+		response.Metadata.FeatureNames.Val = append(response.Metadata.FeatureNames.Val, entityName)
```

Comment:
why it's called `entityName` if it's being added to features list?

Reply:
I'm not sure what the reason behind keeping compatibility with python. Isn't this a complete replacement for python code?

Reply:
> populateEntityRowsFromColumnar
From the code above I see that this function is used only to add request features

Reply:
It actually combines requested entities + entities in ODFV + request features but I reuse the hash map. I wanted to make the code easy to read and understand on the first PR so I tried to mimic python sdk as much as possible - but I'm gonna rewrite more efficiently from now...